process/interceptors/factory/eligibleNodesCache.go

```diff
@@ -0,0 +1,102 @@
+package factory
+
+import (
+	"sync"
+
+	"github.com/multiversx/mx-chain-core-go/core"
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-go/process"
+)
+
+type eligibleNodesCache struct {
+	// eligibleNodesMap holds a map with key shard id and value
+	//		a map with key epoch and value
+	//		a map with key eligible node pk and value struct{}
+	// this is needed because:
+	// 	- a shard node will intercept proof for self shard and meta
+	// 	- a meta node will intercept proofs from all shards
+	eligibleNodesMap    map[uint32]map[uint32]map[string]struct{}
+	mutEligibleNodesMap sync.RWMutex
+
+	peerShardMapper  process.PeerShardMapper
+	nodesCoordinator process.NodesCoordinator
+}
+
+func newEligibleNodesCache(
+	peerShardMapper process.PeerShardMapper,
+	nodesCoordinator process.NodesCoordinator,
+) (*eligibleNodesCache, error) {
+	if check.IfNil(peerShardMapper) {
+		return nil, process.ErrNilPeerShardMapper
+	}
+	if check.IfNil(nodesCoordinator) {
+		return nil, process.ErrNilNodesCoordinator
+	}
+
+	return &eligibleNodesCache{
+		eligibleNodesMap: map[uint32]map[uint32]map[string]struct{}{},
+		peerShardMapper:  peerShardMapper,
+		nodesCoordinator: nodesCoordinator,
+	}, nil
+}
+
+// IsPeerEligible returns true if the provided peer is eligible
+func (cache *eligibleNodesCache) IsPeerEligible(pid core.PeerID, shard uint32, epoch uint32) bool {
+	cache.mutEligibleNodesMap.Lock()
+	defer cache.mutEligibleNodesMap.Unlock()
+
+	var err error
+	eligibleNodesForShard, hasShardCached := cache.eligibleNodesMap[shard]
```

Comment:
can you use instead read lock and only use Lock when you need to update the map?
then most of the accesses could be done without waiting.

Reply:
updated as suggested